### ISSUE TITLE
Trouble getting slamdunk map to work

### ISSUE THREAD

**Issue by FelixKrueger** [t-neumann/slamdunk#37]
Hi Tobias,

I am trying to get SlamDunk to work for the first time, so please excuse my newbie question. I have cloned slamdunk from Github and run ` pip install --user -r ../requirements.txt` afterwards. I next installed NextGenMap on our cluster, and it seems to work:

```
~/slamdunk (master) $ ngm
[MAIN] NextGenMap 0.5.5
[MAIN] Startup : x64 (build Sep 12 2018 14:58:37)
[MAIN] Starting time: 2018-09-14.16:19:57
Usage:  ngm [-c <path>] {-q <reads> [-p] | -1 <mates 1> -2 <mates 2>} -r <reference> -o <output> [parameter]
...
```

When I now try to run slamdunk map, I am getting the following error:
```
 ~/slamdunk/bin/slamdunk map  -ss -r genome/GRCm38.mfa -o SlamDunk_results lane6378_TCCTGAGC_TCTCTCCG_Plate1_H04_L001_R1_val_1.fq.gz
Traceback (most recent call last):
  File "/bi/home/fkrueger/slamdunk/bin/slamdunk", line 28, in <module>
    slamdunk.run()
  File "/bi/home/fkrueger/slamdunk/slamdunk/slamdunk.py", line 414, in run
    mapper.checkNextGenMapVersion()
  File "/bi/home/fkrueger/slamdunk/slamdunk/dunks/mapper.py", line 43, in checkNextGenMapVersion
    raise RuntimeError('Could not get NextGenMap version. Please reinstall slamdunk package.')
RuntimeError: Could not get NextGenMap version. Please reinstall slamdunk package.
```

Any idea what I am doing wrong? Also, if the data is paired-end, is there a way to specify this for SlamDunk or are the reads treated independently of each other? 
Many thanks, Felix

**Comment by t-neumann** [comment]
Hi Felix,

these days, installing everything manually is not really recommended anymore. Looks like it did not find ngm in the appropriate place (slamdunk/contrib).

I would strongly advise you to install slamdunk directly into a conda environment (http://t-neumann.github.io/slamdunk/docs.html#bioconda) or use our Docker container (https://hub.docker.com/r/tobneu/slamdunk) - you can also build a Singularity image from it if required from your HPC environment.

**Comment by FelixKrueger** [comment]
Hi Tobias,
Thanks for the reply. We are running a cluster based on Rocks 6, so we really have to install apps in the apps folder and then load the modules with `module load`. Rocks 6 also doesn't support Docker. I have never used Singularity, I only know that is supposed to be similar to Docker.

It is probably easier to set up a virtual Linux machine on my Desktop so I get control over the installation process, but that means I am trading in hundreds of cores and a few TB of RAM :)

**Comment by t-neumann** [comment]
Hi Felix,

I would first ask your HPC guy whether they have a singularity module installed. Our IT refused Docker for permission reasons, but offered Singularity as a substitute.

If it is available, you can easily build and run a singularity image from our Docker container like this:

```
singularity build slamdunk.simg docker://tobneu/slamdunk:latest
```

This would save you all the installation / dependency troubles.

Alternatively of course, maybe you have an anaconda or miniconda module which you could load to then install slamdunk into a custom environment.

**Comment by FelixKrueger** [comment]
Hi Tobias,

We have now looked a little at Singularity, and eventually installed it on our development server. Your command seemed to do the right thing:

```
singularity build slamdunk.simg docker://tobneu/slamdunk:latest
Docker image path: index.docker.io/tobneu/slamdunk:latest
Cache folder set to /home/fkrueger/.singularity/docker
[6/6] |===================================| 100.0%
Importing: base Singularity environment
Exploding layer: sha256:124c757242f88002a858c23fc79f8262f9587fa30fd92507e586ad074afb42b6.tar.gz
Exploding layer: sha256:9d866f8bde2a0d607a6d17edc0fbd5e00b58306efc2b0a57e0ba72f269e7c6be.tar.gz
Exploding layer: sha256:fa3f2f277e67c5cbbf1dac21dc27111a60d3cd2ef494d94aa1515d3319f2a245.tar.gz
Exploding layer: sha256:398d32b153e84fe343f0c5b07d65e89b05551aae6cb8b3a03bb2b662976eb3b8.tar.gz
Exploding layer: sha256:afde35469481d2bc446d649a7a3d099147bbf7696b66333e76a411686b617ea1.tar.gz
Exploding layer: sha256:ff5b12620374e33da8f0f844932fb5060fb7a4ed8db0b20fad348aceb58743b9.tar.gz
Exploding layer: sha256:fb63b7057c66894076ae5a144342e3ed9793faca25f6f9fa6ec4958167aa03a6.tar.gz
WARNING: Building container as an unprivileged user. If you run this container as root
WARNING: it may be missing some functionality.
Building Singularity image...
Singularity container built: slamdunk.simg
Cleaning up...
```

Trying to run a shell command did this though:
```
 singularity shell slamdunk.simg
Singularity: Invoking an interactive shell within container...

FATAL: kernel too old
```

Running the Hello World image, or 
```
singularity pull --name nf-core-rnaseq.simg docker://nfcore/rnaseq
 singularity shell nf-core-rnaseq.simg
Singularity: Invoking an interactive shell within container...

Singularity nf-core-rnaseq.simg:~>
```

worked fine, so Singularity as such seems to work fine. Any idea what I might be doing wrong? (and sorry I am quite new to this).

**Comment by t-neumann** [comment]
Hi,

which singularity version are you using?

An what happens if you execute slamdunk like this?

```
singularity exec slamdunk.simg slamdunk -h
```

**Comment by FelixKrueger** [comment]
Hi,
```
singularity --version
2.6.0-dist
```
and
```
singularity exec slamdunk.simg slamdunk -h
FATAL: kernel too old
```

**Comment by t-neumann** [comment]
What is the setup on your deployment server?

**Comment by t-neumann** [comment]
Indeed, but apparently there is still some issue. I have collaborators using it without issues on different systems in Italy and the UK, so it can't be a general container issue.

We are using an older Singularity version - is it possible to try to downgrade?

singularity/2.4.5

**Comment by FelixKrueger** [comment]
We just downgraded Singularity to 2.4.5, but all errors are exactly the same.

**Comment by FelixKrueger** [comment]
Maybe this is relevant (according to the Singularity FAQ):
```
Are Singularity containers kernel dependent?
No, never. But sometimes yes
…

There is also some level of glibc forward compatibility that must be taken into consideration for any container system. For example, I can take a Centos-5 container and run it on Centos-7, but I can not take a Centos-7 container and run it on Centos-5.
```

**Comment by t-neumann** [comment]
Hm I read about that glibc issue in a different context.

But I'm pretty sure it's not a container issue, since you pull the Docker container build from Dockerhub and from that build your own singularity container on your machine.

But maybe indeed is actually a problem of the OS. I don't suppose you could run it on such a system?

```
LSB Version:   	:core-4.1-amd64:core-4.1-noarch
Distributor ID:	CentOS
Description:   	CentOS Linux release 7.4.1708 (Core)
Release:       	7.4.1708
Codename:      	Core
```

**Comment by FelixKrueger** [comment]
Hi Tobias,

I'm afraid we can't change the OS of our (900 core) cluster easily. I think I am going to experiment with building a basic singularity image with an different operating system, and then trying to install SlamDunk with Conda on there. Do you think that would work?

**Comment by t-neumann** [comment]
Yes installing slamdunk in a conda environment should definitely work.
Can't you just use it without running it within a Singularity image? As long as you have conda on your  development server installed, it should definitely work.

**Comment by FelixKrueger** [comment]
We would really like to get it going with Singularity because the end goal would be to be able to deploy it on our cluster, and it seems like we can run Singularity there, but not conda...

I have been making progress today, I can get a singularity image to install miniconda, SlamDunk and NextGenMap (https://github.com/FelixKrueger/SlamDunk_Shub/blob/master/Singularity and https://www.singularity-hub.org/containers/4919) and build alright, but when I pull it down as `singularity pull shub://FelixKrueger/SlamDunk_Shub:latest` I can run a shell, but can't find SlamDunk (or conda or ngm) on there. Probably just me not understanding how Singularity works...

I have in parallel created a Centos 7 based VirtualMachine (with Virtualbox), installed pretty much exactly the same as in the Singularity image, and everything, including `slamdunk` and `ngm` works on the VM. So I think really all I need is to 'activate' or find the programs in the Singularity image, and Bob's my uncle. Would you know off hand how to do that? Have a good weekend, Felix

**Comment by FelixKrueger** [comment]
I think one problem was that I installed the programs to $HOME, which was then masked when my host home directory was overlaid to the container one. I have now changed it so that Conda gets installed into /opt on the container, but I still can't really find where SlamDunk is installed with the create command. I have just joined the Singularity Slack channel and am hoping that some of the folks over there will be able to help me out. I think we are getting close, I'll update once I manage to get it going. Thanks for your help. Felix

(I am quite lucky that the researchers who would like their SLAM-Seq data analysed are very patient :) )

**Comment by t-neumann** [comment]
Hi Felix,

good to hear you're finally able to run it. The 3' UTR step won't do any harm - worst case it won't recover multimappers which you would not have had in the first place without the filtering.
Stefan published some 3' end annotation along with his paper based on mESCs, but we use the plain 3' UTR coordinates you can obtain from UCSC for mouse and human samples.

You are of course welcome to contact me again if you run into further questions and I will make sure to send your regards to Ami :)

Best,

Tobi
